operator-framework/operator-lifecycle-manager
# Prevent OLM from creating `InstallPlan`s when bundle unpack fails

Prerequisites:
* https://github.com/operator-framework/api/pull/282
* https://github.com/operator-framework/operator-lifecycle-manager/pull/2954
* https://github.com/operator-framework/operator-lifecycle-manager/pull/2952

---



**Description of the change:**

This PR changes how `InstallPlan`s are being created.

**Motivation for the change:**

Today OLM creates InstallPlans even if bundle images are not available for some reason. For example, catalog source contains references to an unreachable registry.

Consider this scenario:
* User installs an operator and enables auto upgrades
* Catalog source receives an update which references unreachable bundle images (e.g. pre-release from a private registry)
* OLM creates an `InstallPlan` for an upgrade which features unreachable bundle images
* At this point upgrade is stuck since the cluster can not pull bundle images
* Catalog source receives a corrective update which contains only reachable bundle images

Workaround: to resolve this issue users have to manually remove a failed `InstallPlan`. After that OLM is expected to create a new `InstallPlan` based on the latest state of catalog sources.

The goal: We would like to find a solution which prevents `InstallPlan` from being created with unreachable images in the first place. This eliminates the need to manually delete `InstallPlan` in scenarios like the one described above.

**Architectural changes:**



Before the change:

* On a `Namespace` reconciliation (`syncResolvingNamespace`) OLM creates an `InstallPlan` if required (e.g. new sub is created or a new update in a catalog is available for an existing sub)
* On `InstallPlan` reconciliation (`syncInstallPlans`) OLM creates a `Job` to unpack a relevant bundle.
  * On unsuccessful job unpacking OLM reports failures as `InstallPlan` conditions

After this change:

* On a `Namespace` reconciliation (`syncResolvingNamespace`):
  * OLM first creates a `Job` to unpack a relevant bundle and waits for the job to sucessfully complete
    * On unsuccessful job unpacking OLM reports failures as `Subscription` conditions (set on all subs in the namespace)
  * Only after successfull completion OLM creates an `InstallPlan`



**Testing remarks:**

To reproduce:

* Create a test cluster and install OLM (`make run-local` from this repo can be helpful)
* Create a catalog source:
    ```yaml
    apiVersion: operators.coreos.com/v1alpha1
    kind: CatalogSource
    metadata:
      name: alex-operators
      namespace: olm
    spec:
      image: quay.io/agreene/index:unreachable-bundle-upgrade
      displayName: Alex Operators
      priority: -100
      publisher: Alex
      sourceType: grpc
      updateStrategy:
        registryPoll:
          interval: 10m0s
    ```
* And create a subscription:
    ```yaml
    apiVersion: operators.coreos.com/v1alpha1
    kind: Subscription
    metadata:
      name: quay
      namespace: operators
    spec:
      channel: unavailable-image-upgrade
      installPlanApproval: Automatic
      name: quay
      source: alex-operators
      sourceNamespace: olm
      startingCSV: quay-operator.v3.8.3
    ```

In master - you will see two `InstallPlan`s: one for `quay-operator.v3.8.3` (successfull install) and one for `quay-operator.v3.8.4` which will eventually (10min by default) report failure to unpack in bundle lookup conditions.

In this branch you will see only one `InstallPlan` for `quay-operator.v3.8.3` (successfull install). Eventually `Subscription` will get a condition reporting failure to unpack (`SubscriptionBundleUnpackFailed`). During unpacking it should report `UnpackingInProgress`.

**Reviewer Checklist**
- [ ] Implementation matches the proposed design, or proposal is updated to match implementation
- [ ] Sufficient unit test coverage
- [ ] Sufficient end-to-end test coverage
- [ ] Bug fixes are accompanied by regression test(s)
- [ ] e2e tests and flake fixes are accompanied evidence of flake testing, e.g. executing the test 100(0) times
- [ ] tech debt/todo is accompanied by issue link(s) in comments in the surrounding code
- [ ] Tests are comprehensible, e.g. Ginkgo DSL is being used appropriately
- [ ] Docs updated or added to `/doc`
- [ ] Commit messages sensible and descriptive
- [ ] Tests marked as `[FLAKE]` are truly flaky and have an issue
